expose addContact method

Wo1v3r opened the pull request.

Wo1v3r commented:
@rangoldberg @omrital 

This PR adds a `addContact(contact)` function both to iOS and android.
Tested both on simulators

I see that on Android you guys already created classes for the structured fields of contacts, I'd love to refactor the huge chunk of code I added there to use it but my IDE was kind of broken.

Would be amazing if this could be merged so we can start using it for our use case and I'll refactor that later on if you want.

LMK if there are any changes needed.

Wo1v3r commented:
@itzikkit @bruchim

rangoldberg commented:
probably meant @gran33

gran33 commented on Example/package.json:
```diff
@@ -3,7 +3,7 @@
 	"private": true,
 	"scripts": {
-		"start": "node node_modules/react-native/local-cli/cli.js start",
+    "start": "node node_modules/react-native/local-cli/cli.js start",
```
plz don't PR lint

gran33 commented on Example/src/demo-app-component.js:
```diff
@@ -1,92 +1,110 @@
       loading: true,
       actualCount: 0,
-      count: 'unknown'
+      count: 'unknown',
```
plz remove

gran33 commented on Example/src/demo-app-component.js:
```diff
@@ -1,92 +1,110 @@
       const count = await this.pagedContacts.getContactsCount();
       this.setState({
-        count: count
+        count: count,
```
plz remove

gran33 commented on Example/src/demo-app-component.js:
```diff
@@ -1,92 +1,110 @@
+        );
         contacts.push(...newContacts);
-        this.setState({actualCount: contacts.length});
+        this.setState({ actualCount: contacts.length });
```
plz don't PR lint

gran33 commented on Example/src/demo-app-component.js:
```diff
@@ -1,92 +1,110 @@
+
   render() {
-    const {data} = this.state;
+    const { data } = this.state;
```
plz don't PR lint

gran33 commented on Example/src/demo-app-component.js:
```diff
@@ -1,92 +1,110 @@
 
   renderList() {
-    const {data, count, actualCount} = this.state;
+    const { data, count, actualCount } = this.state;
```
plz don't PR lint

gran33 commented on Example/src/demo-app-component.js:
```diff
@@ -1,92 +1,110 @@
         <FlatList
           data={data}
-          renderItem={({item}) => <Text>{item.label}</Text>}
+          renderItem={({ item }) => <Text>{item.label}</Text>}
```
plz don't PR lint

gran33 commented on Example/src/demo-app-component.js:
```diff
@@ -1,92 +1,110 @@
 
   renderLoading() {
-    const {loading, animationConfig, actualCount, count} = this.state;
+    const { loading, animationConfig, actualCount, count } = this.state;
```
plz don't PR lint

gran33 commented on Example/src/demo-app-component.js:
```diff
@@ -1,92 +1,110 @@
+            overlay
+            {...animationConfig}
+          />
+        )}
```
plz don't PR lint

gran33 commented:
@Wo1v3r plz remove unrelated files and lines from the PR

Wo1v3r commented:
@gran33  I updated the PR, removed the changes from lint and unnecessary imports

Wo1v3r commented on android/src/main/java/com/wix/pagedcontacts/contacts/ContactsProvider.java:
```diff
@@ -140,4 +154,232 @@ private Cursor queryContacts(QueryParams params) {
+    }
+
+    private int mapStringToPhoneType(String label) {
+        int phoneType;
```
They are similar to the contacts returned by `react-native-paged-contacts` (Same fields mappings, but the other way around) do we really want to expose it separately?

only difference I made is `imageUrl` because I explicitly use a URL for the contact's image.

Wo1v3r commented on index.js:
```diff
@@ -120,6 +121,29 @@ export class PagedContacts {
+   */
+  addContact(contact) {
+    if (Platform.OS === 'android') {
+      await PermissionsAndroid.request(
```
it will throw an error that should be caught by whoever is using this function. similar to `requestAccess` above.

bruchim replied:
cool

Wo1v3r commented on android/src/main/java/com/wix/pagedcontacts/contacts/ContactsProvider.java:
```diff
@@ -140,4 +154,232 @@ private Cursor queryContacts(QueryParams params) {
         return ids;
     }
+
+    public void saveContact(ReadableMap contact, Promise promise) {
```
Alright, I'll move the try catch to the exposed ReactMethod and add a message to the rejected promise. WDYT ? @bruchim

bruchim replied:
great

Wo1v3r commented:
@bruchim  I refactored the android code and added some missing fields to iOS.

Now the field keys are the same as the ones exposed by the module, I tested using the same JSON on both platforms too. Some fields that are not in use are ones that do not exist in both platforms ofcourse.

Wo1v3r commented on android/src/main/AndroidManifest.xml:
```diff
@@ -2,5 +2,6 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
           package="com.wix.pagedcontacts">
     <uses-permission android:name="android.permission.READ_CONTACTS" />
+    <uses-permission android:name="android.permission.WRITE_CONTACTS" />
```
@gran33  just note this

gran33 replied:
I'll bump major, it will be 3.0.0
